pkg/proc/target.go

```diff
@@ -1,28 +1,119 @@
 package proc
 
+import (
+	"errors"
+	"go/ast"
+)
+
+// ErrDirChange is returned when trying to change execution direction
+// while there are still internal breakpoints set.
+var ErrDirChange = errors.New("direction change with internal breakpoints")
+
 // Target represents the process being debugged.
 type Target struct {
 	Process
 
+	// Breakpoint table, holds information on breakpoints.
+	// Maps instruction address to Breakpoint struct.
+	breakpoints BreakpointMap
+
 	// fncallForG stores a mapping of current active function calls.
 	fncallForG map[int]*callInjection
 
 	// gcache is a cache for Goroutines that we
 	// have read and parsed from the targets memory.
 	// This must be cleared whenever the target is resumed.
 	gcache goroutineCache
+
+	// threadToBreakpoint maps threads to the breakpoint that they
+	// have were trapped on.
+	threadToBreakpoint map[int]*BreakpointState
```

Comment:
I don't like that a chunk of the thread state now goes into a map instead of being in its logical place (inside the thread object).

Reply:
Honestly I could go either way on this.

My thoughts are removing it from the Thread object(s) has a few benefits:

1. Remove redundancy (not a huge win overall, I admit)
2. We keep better track of the state this way. Instead of always iterating every thread and asking it if it's stopped at a breakpoint, we have a definitive mapping right here.
3. Easier to cleanup and manage; this is ephemeral state, not necessarily an actual piece of information on an operating system thread. We can easily cleanup this state by wiping it out and creating a new map.
4. The backend(s) no longer care about breakpoint state for a thread, so we'd be storing this information in the backend solely for the use of the consumer, even though the consumer itself (`Target`) is using and managing the information.

Reply:
Responding point to point:

1. Agree, not much of a win given that it's just a field and a getter
2. There are two places in this patch where Target.threadToBreakpoint is iterated, by contrast there are four places where we have a Thread and we want to get its breakpoint
3. Of the two places where we iterate threadToBreakpoint one is the thread resume logic where we do not want to just wipe the map because each thread needs to be single stepped, so in 50% of the cases the easy cleanup is not a win
4. In principle, I'd rather the backend knew. Even if it doesn't need to know now it might in the future and there's no reason to erect artificial barriers to it

Reply:
I agree with your points, there is one more things to consider though. Once all backends are in `pkg/internal`, we will have to initialize them via `Target`, which means `pkg/proc` will be importing the backends instead of the other way around.

If both `pkg/proc` and the various backends know about breakpoints we get into a cyclic dependency issue.

Reply:
@aarzilli I think this is the only outstanding item of contention.

Reply:
Either a third package will initialize them or the initialization will keep happening as it does now and their path won't have 'internal' in it. What does putting them into an internal package get us?

At the moment proc can call anything in a backend and the backend can call anything in proc. I'd like that to continue to be true.

Reply:
The backends aren't safe to use by themselves, so I'd prefer if nothing outside of proc is able to import them.

Reply:
In any event, we can go back and forth on this point at a later time in the refactor. I'm working on reverting this so "current breakpoint" state stays on threads for now.

Reply:
> The backends aren't safe to use by themselves, so I'd prefer if nothing outside of proc is able to import them

If you look at what the backends actually export right now, almost everything you can call can be safely called by anyone, because the only things you can call are the launch and attach functions, which return Target objects. The exception here are the Ptrace* functions in the native package that are exported for no reason at all.

Reply:
They also all export the Process and Thread concrete types but there's no way to create them and we could very easily un-export them.

Reply:
@aarzilli I agree that we should likely un-export the Process and Thread types in the backends, and anything else that doesn't need to be exported. I guess my position on it is, we don't expect or want anything outside of proc to import the backends directly, so why not enforce that using an internal import path? Right now the only reason why the backends are being imported is, as you mentioned, the Launch, Attach, etc functions. To me though that feels backwards as well, that was something I relented on during initial refactor discussions but still feel can be done better. I don't think the backends returning `proc.Target` is the right thing. I still prefer the more generic Launch / Attach methods that were in the original refactor sketch, something like `proc.Launch` instead of the debugger package making the determination on which backend to use. 

Ideally I see it as:

1. Debugger (or test code) wants to start a process
2. Debugger (or test code) asks proc package to Launch or Attach
3. proc package handles figuring out which backend to use and initialize it
4 proc package returns its own `proc.Target` object, which holds an instance of the selected backend.

This ends up looking like:

```
debugger --> proc.target --> { native, gdbserial, core }
```

Instead of:

```
debugger --> { native, gdbserial, core } --> proc.target
```

Either way, the only relevance that change has to this current patch is handling the cyclic dependency problem. If we wait on that we can discuss and solve this at another time.

I regret that I didn't first draft a design document to discuss the scope of this refactor and the desired end result. I think that would have helped guide the discussion a lot better. It might still be helpful, now that I think about it.

Reply:
> I guess my position on it is, we don't expect or want anything outside of proc to import the backends directly, so why not enforce that using an internal import path? Right now the only reason why the backends are being imported is, as you mentioned, the Launch, Attach, etc functions. To me though that feels backwards as well

I don't agree with this. Yes, we don't expect anything outside of *debugger* to import the backends, we also do not expect anything outside of debugger to import proc itself. And we don't expect anything outside of service to import debugger. 

And I don't think having the constructors in the backend is backwards: you want to use the native backend, you import the native backend. If having the switch statement duplicated in debugger and proc_test we can use a mechanism like database/sql with the init function of the backends calling proc.Register.

Reply:
> calling proc.Register

It does occour to me now that proc.Register is already take, but you get the point.

Also, I'm not in principle opposed to proc.Launch/proc.Attach it's just that I don't see it as a big win and having the backends be unable to see inside pkg/proc is too great of a downside.

Reply:
> > I guess my position on it is, we don't expect or want anything outside of proc to import the backends directly, so why not enforce that using an internal import path? Right now the only reason why the backends are being imported is, as you mentioned, the Launch, Attach, etc functions. To me though that feels backwards as well
> 
> I don't agree with this. Yes, we don't expect anything outside of _debugger_ to import the backends, we also do not expect anything outside of debugger to import proc itself. And we don't expect anything outside of service to import debugger.

I should have been more clear, I was phrasing that based on future work, if we were to invert the initialization (using only proc, not the various backends). As of now, you're correct, nothing except `debugger` (and tests) should be importing the backends.